core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -647,7 +648,7 @@ public boolean putNewKey(Object keyBase, long keyOffset, int keyLength,
       assert(bitset != null);
       assert(longArray != null);
 
-      if (numElements == MAX_CAPACITY || !canGrowArray) {
+      if (numElements == MAX_CAPACITY || numElements > growthThreshold && !canGrowArray) {
```

Comment:
When we have `numElements  <= growthThreshold && !canGrowArray`, it is guaranteed that our page still has space to put this key, right?